### ISSUE TITLE
TestCaseSource properties queried before SetUpFixture SetUp method is called

### ISSUE THREAD

**Issue by baracek** [nunit/nunit#141]
My scenario requires some dynamic loading before the TestCaseSource is called which changes what is returned from the TestCaseSource property.

An example would be:

``` C#
[TestFixture]
public class SampleTest
{
   public IEnumerable<int> AllTestValues
   {
      get
      {
         // testsourcevalues is null here
         foreach ( int testValue in SetupSystem.testSourceValues )
         {
            yield return testValue;
         }
      }
   }

   [Test, TestCaseSource( "AllTestValues" )]
   public void SampleTestCase( int testValue )
   {    
   }
}

[SetupFixture]
public class SetupSystem
{
   public static List<int> testSourceValues;
   [Setup]
   public void SetUp()
   {
      testSourceValues = new List<int>();
      testSourceValues.Add( 1 );
      testSourceValues.Add( 2 );
      testSourceValues.Add( 3 );
   }
}
```

**Comment by beloquintana** [comment]
I have a similar issue:

In my scenario I need to change the data directory path of my tests. 
I am using --params|p=PARAMETER to do this. I am setting a new path directory by parameters in a method with the annotation "OneTimeSetUp".

If Nunit runs first TestCaseSource, I can't resolve it with a good solution. In my opinion correct behavior should be to run "setup" first, then load and create test cases.

**Comment by CharliePoole** [comment]
It's not so much a question of "correct" behavior as it is of "defined" behavior. NUnit has a specified sequence of loading and executing tests...

* Loading
  * Find all the test methods
  * Determine parameters to use (may involve invoking user code, but not test code)
  * Create the internal representation of the tests
* Running (Note that this may happen more than once per test load)
  * Run OneTimeSetUps starting at the assembly level and proceeding downward.
    * Run SetUp for test
    * Run the test
    * Run TearDown for test
    * Repeat for each test
  * Run OneTimeTearDowns starting at the test fixture and proceeding upward.

As you can see, the loading of the tests, including invoking any TestCaseSource is not only before setup, but is __way__ before it, in an entirely different phase. The separation between Loading and Running is baked in. Your TestCaseSource is actually not part of your test, it's something that helps NUnit to create your tests. It is invoked, for example, if you run with the `--explore` option or if the Gui simply wants to load the tests for display.

This issue is about an __enhancement__ to NUnit that would allow the tests to be discovered dynamically (and possibly differently) each time you run them. With the new feature, there could be an entirely different set of tests - maybe even a different number of tests - and GUIs would not be able to display them until __after__ they had executed.

This is expected to be a very useful feature when we do it, but it's definitely an advanced one. It usually is not needed by folks having problems such as yours. What is a bit different in your situation is your use of command-line parameters.

Command-line parameters are accessed through the TestContext. However, there is no TestContext when tests are being loaded, only when they are created. It might be useful to you if we provided some way to access command-line paramters from a TestCaseSource method at load time. If you think so, please create a feature request.

**Comment by beloquintana** [comment]
I think that can be usefull to have a listener or something before Loading. I was able to solve it using TestContext after running TestCaseSource and before I get my data path. I was able to use TestContext in the method that provide data to my TestCaseSource and it's working fine.

Thanks.

**Comment by CharliePoole** [comment]
@beloquintana You're right... you can access most static members of the TestContext at load time... I had forgotten we did that.

**Comment by bronumski** [comment]
Wondering if there is any movement on this and if I can be of any assistance from a use case point of view?

I currently have a scenario where I run a suit of selenium tests however these tests are targeted at different instances with different types of data. Currently prior to the test run a number of static classes are instantiated which pulls in data from a rest endpoint via a httpclient and this data is then used to generate the test cases for that instance. This works fine but there is a disconnect between some of the wire up and to developers coming into the project a lot of it feels like magic. It also has a knock on effect to anything that has to happen upstream of the api calls also having to be done inside statics.

**Comment by bdonnahue** [comment]
I am also wondering. I opened https://github.com/nunit/nunit/issues/2869 as a related issue

**Comment by ravensorb** [comment]
Ditto -- its been almost 5 years and this is still open.  With DI/IoC, dynamic objects, policy based coding, design by contract, etc all becoming main stream there really is a huge need for the ability to 'setup' the app env before loading the test data.  Any chance we can get this on a solid roadmap?

**Comment by CharliePoole** [comment]
Even longer... see issues #3 and #7, each of which links to the original Launchpad issue.

The plan for NUnit 3 was to have both static and dynamic generation of test case data, where "static" means what we have already and "dynamic" is basically what you are asking for.

Over time, various people have chimed in to suggest we __change__ the existing behavior rather than adding a new kind of data source. I think that's a bad idea, since we would have to figure out how to change it in a way that didn't break anyone. Also, there will always be some use for the arguably more efficient static approach.

Remember that we don't have a mechanism for creating a plan and making people follow it like commercial enterprises can do. :smile: Consequently, we don't "assign" work on this project, except to ourselves as individuals. 

This issue requires somebody who has both the skill level, knowledge of how NUnit works and motivation to take on the task. So far, nobody has popped up to do it. :disappointed:

**Comment by shatulsky** [comment]
You can workaround it like this:

```
private static IEnumerable<TestCaseData> GetTestData()
{
   yield return new TestCaseData(
   new Lazy<Model>(() => *generate data*)
   {
      TestName = "TC 1"
   };

   yield return new TestCaseData(
   new Lazy<Model>(() => *generate data*)
   {
      TestName = "TC 2"
   };
}

[TestCaseSource(nameof(TestCaseData))]
public void Test1(Lazy<Model> lazyMessage)
{
}
```

**Comment by CharliePoole** [comment]
@baracek 

That would work but without your use of `TestName` all the tests would have the same name. Obviously you understand that but I wanted to point it out for others. Anyone using this approach widely will have to give unique names to every test case, rather than relying on NUnit to do it.

**Comment by gnaneswarreddy** [comment]
I have the similar setup as mentioned below, but my generate data comes from excel where in the Test1 i need only two arguments(username and password) this gives error, but in the second test I need 4 arguments which works fine.
private static IEnumerable<TestCaseData> GetTestData()
{
 return Framework.ExcelDataReader
          .ReadExcel(Excelfilename,Excelfiletab)
          .Select(c => new TestCaseData(c[0], c[3], c[6], c[12]))
}

[TestCaseSource(nameof(GetTestData))]
public void Test1(string c[0],c[3])        //This is giving error, if I give 2 arguments it is giving error(Is there any way i can use only
                                                                2 Arguments as i dont need other two arguments
{
}

[TestCaseSource(nameof(GetTestData))]
public void Test2(string c[0], c[3], c[6], c[12])  //This works fine without any issue
{
}

**Comment by mikkelbu** [comment]
@gnaneswarreddy I don't think this is related to the problem in this issue. If you want to execute parametric tests that expect different input, then you will need to create two separate TestCaseSources or generalise the source and the test to work on arrays and then extract the elements inside the tests.
